pkg/store/cache/caching_bucket.go

```diff
@@ -26,92 +29,178 @@ import (
+	opConfig   interface{}
+}
+
+// Operation-specific configs.
+type iterConfig time.Duration
+
+type existsConfig struct {
+	existsTTL, doesntExistTTL time.Duration
+}
 
-	requestedChunkBytes prometheus.Counter
-	fetchedChunkBytes   *prometheus.CounterVec
-	refetchedChunkBytes *prometheus.CounterVec
+type getConfig struct {
+	contentTTL, existsTTL, doesntExistTTL time.Duration
+}
 
-	objectSizeRequests prometheus.Counter
-	objectSizeHits     prometheus.Counter
+type getRangeConfig struct {
+	subrangeSize               int64
+	objectSizeTTL, subrangeTTL time.Duration
+	maxSubRequests             int
 }
 
-func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+type objectSizeCfg time.Duration
+
+// NewCachingBucket creates caching bucket with no configuration. Various "Cache*" methods configure
+// this bucket to cache results of individual bucket methods. Configuration must be set before
+// caching bucket is used by other objects.
+func NewCachingBucket(b objstore.Bucket, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
 	if b == nil {
 		return nil, errors.New("bucket is nil")
 	}
-	if c == nil {
-		return nil, errors.New("cache is nil")
-	}
 
 	cb := &CachingBucket{
 		Bucket: b,
-		config: chunks,
-		cache:  c,
 		logger: logger,
 
-		requestedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_requested_chunk_bytes_total",
-			Help: "Total number of requested bytes for chunk data.",
-		}),
-		fetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_fetched_chunk_bytes_total",
-			Help: "Total number of fetched chunk bytes. Data from bucket is then stored to cache.",
-		}, []string{"origin"}),
-		refetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_refetched_chunk_bytes_total",
-			Help: "Total number of chunk bytes re-fetched from storage, despite being in cache already.",
-		}, []string{"origin"}),
-		objectSizeRequests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_objectsize_requests_total",
-			Help: "Number of object size requests for objects.",
-		}),
-		objectSizeHits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_objectsize_hits_total",
-			Help: "Number of object size hits for objects.",
-		}),
-	}
-
-	cb.fetchedChunkBytes.WithLabelValues(originBucket)
-	cb.fetchedChunkBytes.WithLabelValues(originCache)
-	cb.refetchedChunkBytes.WithLabelValues(originCache)
+		operationConfigs: map[string][]*operationConfig{},
+
+		requestedGetRangeBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_getrange_requested_bytes_total",
+			Help: "Total number of bytes requested via GetRange.",
+		}, []string{config}),
```

Comment:
what is config here.. you mean "config"?

Reply:
There was `const config="config"` on the top, but it's now inlined.

Reply:
ah thanks XD it confused me